scene/resources/curve.cpp

```diff
@@ -2251,6 +2287,7 @@ void Curve3D::_bind_methods() {
 	ClassDB::bind_method(D_METHOD("get_point_in", "idx"), &Curve3D::get_point_in);
 	ClassDB::bind_method(D_METHOD("set_point_out", "idx", "position"), &Curve3D::set_point_out);
 	ClassDB::bind_method(D_METHOD("get_point_out", "idx"), &Curve3D::get_point_out);
+	ClassDB::bind_method(D_METHOD("get_point_baked_posture", "idx", "apply_tilt"), &Curve3D::get_point_baked_posture, DEFVAL(false));
```

Comment:
```suggestion
```

Do not bind posture as it should be an internal concept in the curve3D stage (before path3D).

Reply:
This can be implemented externally enough by simply putting `compose_posture()` functions in public as described below, so you need implement this in EditorPlugin side.

Reply:
the main problem is with the mapping between control points to baked points. see the thread bellow.

Reply:
It would be most memory saving to create this cache array as a local variable in Curve3D::_bake(), and pass it to Path3DEditor via the signal argument, and let the Editor cache it by duplicating it. Then, you would need to run `_bake()` each time you select the curve. Also, the array calculation and signal should only be fired with `#ifdef TOOLS_ENABLED`.

Reply:
Then, `get_point_baked_posture()` may need to be reimplemented, but that is true for Path3DEditor as well as other 3DGizmos.
For example, there are several requests to use 3DTransformGizmo in the game, but to do so, it would have to be a tools build (Godot core itself in Game) or the user would have to reimplement the Gizmo.
At least if you make it public, you can use it as a custom_module.

Reply:
> It would be most memory saving to create this cache array as a local variable in Curve3D::_bake(), and pass it to Path3DEditor via the signal argument 

It seems we agree that this cache array needs to be generated by `Curve3D`, but have different opinions about where it should be stored
- on `Curve3D` instance itself
- or on `Path3DEditor`

## Memory saving is negligible

The argument for the second option is saving memory. That is true, but we can show the saving is negligible. 

Let's say we have a curve of 10 control points, then baking will generate 100 baked points(very conservative estimation, could be 1000s or more ). To store these baked points we already have the following vectors on `Curve3D` instance
```
	mutable PackedVector3Array baked_point_cache;
	mutable Vector<real_t> baked_tilt_cache;
	mutable PackedVector3Array baked_up_vector_cache;
	mutable PackedVector3Array baked_forward_vector_cache;
	mutable Vector<real_t> baked_dist_cache;
```
that's (3+1+3+3+1) * 4 * 100 = 4400 bytes.

On the other hand this `points_in_cache` is just 1 integer for each control point, or 4 * 10 =40 bytes. Which means to store it on `Path3DEditor` only saves **40 / 4400 = 0.9%**.  

( If we really want to optimize memory usage, we can remove `baked_tilt_cache` and `baked_forward_vector_cache`, calculating their value one-the-fly, which saves **40%**. But this requires another PR with careful profiling. )

## Create practical and conceptual problems

### Practical problems

First, Storing `points_in_cache` on  `Path3DEditor` requires complex signal plumbing. As you said, we need
- create a new signal `curve_baked` for `Curve3D`
- connect this signal of a `Curve3D` to the `Path3DEditor` when the `Path3D` instance owing the curve is selected.(also unconnect when deselected.)
- trigger `Curve3D::_bake()` every time the `Path3D` is selected. (This is problematic in itself, as the _bake() is expensive, it should only be triggered when curve is "dirty". Simply select it should not make a curve "dirty" )

### Conceptual problems

Second, after careful consideration, I find this is conceptually problematic.

>  It is fine to remove the underscore like compose_posture() and put it in public so that EditorPlugin can use it

It exposes the access to baked points **directly** to end users, but we probably should not. Because baked points are internal and ephemeral, an implementation detail(e.g. we can implement curve without baked points or different way of baking points.). 

That is why the public api only use the baked points **indirectly**. For example
```
	Vector3 sample_baked(real_t p_offset, bool p_cubic = false) const;
```
It refer the position on curve with `p_offset` instead of the index of baked point. Rebaking curves will not impact the end user.

On the other hand, control points are integral part of a curve. It is OK to refer control point directly, hence the `Curve3D::get_point_baked_rotation(int control_point_idx)` method is more conceptually consistent.

For these reasons, I strongly favor the first option, storing `points_in_cache` on the `Curve3D` instance.

Reply:
Other caches are needed for use by other functions in Curve3D, such as those used by Path3D, but points_in_cache is not needed under normal circumstances, and there is no doubt that it is wasteful to keep it in Curve3D.

Reply:
> points_in_cache is not needed under normal circumstances

It is needed when user want to e.g. implement their own in game editor.

 > it is wasteful

Way less than 1%, not that much.

> trigger Curve3D::_bake() every time the Path3D is selected. (This is problematic in itself, as the _bake() is expensive, it should only be triggered when curve is "dirty". Simply select it should not make a curve "dirty" )

This seems much more wasteful

Reply:
> It may be a bit expensive to have _bake() triggered on every selection, but in that case the cleanest way is to use #ifdef TOOLS_ENABLED to enable the cache only in tools builds (FYI, you can refer to the bezier_track HandleMode implementation as similar use case). This means that you will not be able to access its point map in-game, but this is not a problem for normal use cases.

You may want to access it to implement an editor in your game, but since that is an advanced use case as described above, you will usually only need to include it in the tools build.

As a compromise, I would suggest that to make it easier to access it as a custom module, you could use the virtual method and pass an array as pointer to `_bake()` to enable cache usage only when overridden in the tools build or in the custom module like following:

.h
```cpp
virtual void _bake();

void _bake_main(Vector<size_t> *r_arr = nullptr); // Current _bake();
```

.cpp
```cpp
void Curve3D::_bake() {
#ifdef TOOLS_ENABLED
_bake_main(baked_point_cache); // In tools, 
#else
_bake_main(nullptr); // In game, there is no cache array.
#endif
}

void Curve3D::_bake_main(Vector<size_t> *r_arr) {
	if (r_arr) {
		// Process cache array.
	}
	...
}

```

Reply:
> (FYI, you can refer to the bezier_track HandleMode implementation as similar use case)

Do you mean this code?
```
	struct BezierKey {
		Vector2 in_handle; //relative (x always <0)
		Vector2 out_handle; //relative (x always >0)
		real_t value = 0.0;
#ifdef TOOLS_ENABLED
		HandleMode handle_mode = HANDLE_MODE_FREE;
#endif // TOOLS_ENABLED
	};
```

Here I understand and agree with your choice to put `handle_mode` behind `#ifdef TOOLS_ENABLED` gate. I can think two reasons
- without `handle_mode` it saves about 3% of memory for tens of thousands of `BezierKey` in a moderately animated project. 
- the end users are very unlikely to implement their own animation track editor in game.

But here for the `points_in_cache` the situation is different
- without `points_in_cache` it any sames far less than 1% of memory for several dozens of `Curve3D`
- the end user who develops curve heavy games often needs to implement their own in game curve editor, which benefits from easy access to information on control points. (e.g. this godot game https://twitter.com/MightofMe/status/1681650102597722113?s=20 it's in 2D, but the idea is similar in 3D)

> here is no doubt that it is wasteful to keep it in Curve3D.
Design is all about trade-offs. For example, currently each `Curve3D` instance have two caches `baked_tilt_cache` and `baked_forward_vector_cache` that constitute 40% of memory foot print. They are not strict necessary, as they can be calculated on-the-fly. But the current implementation pay those 40% price for simplicity of code and perhaps performance.

Based on the analysis from previous segment, I believe it is a good trade-off to store `points_in_cache` on `Curve3D` instance, paying less than 1% memory foot print, but get benefits of three fold:
- simplicity of logic. No need to add new signal and do all the signal plumbing. No special case of `#ifdef TOOLS_ENABLED` gate
- Performance. No baking when the curve is not "dirty".
- Ease of use for end user. just call the method get the return value, no need of signal or call back.

I would implore you to give these reasoning a thought. As I also means the best for the Godot project. If you find it is hard to decide, is it possible to get a third opinion from another reviewer.

Reply:
Your comparison is not reasonable since the additional signal is not necessary when using #ifdef.

No use case other than implementing an editor is presented, and we do not believe that normal users accessing this cache is a normal use case.

This wastes memory in the normal use case, since the cache would grow by the number of curves if there were a large number of curves. Since Godot tends to prioritize common use cases, it determines that this should only be accessed in tools builds.

For advanced use cases, custom modules are required, not only in Curve3D, but in several other areas as well.

Reply:
> Your comparison is not make sense since the additional signal and calling more _bake() are not necessary when using #ifdef.

I am talking about exactly the tool mode. you need additional signal an call more _bake(). That's complexity and performance loss I am talking about.

> Also the ifdef is only recognized at build the core, so the API does not change for the end user and the API is never complicated.

I am talking about user who want to implement in-game-editor. When they want to get information at a control point, they have to use signal or callback function, instead of a simple method call.

> No use case other than implementing an editor is presented, and we do not believe that normal users accessing this cache is a normal use case. 

Yes, that is only one use case currently I can think of. But it is a genuine use case not some hypothetical one. What is "normal" is highly subjective.

> Moreover, even in a use case with an in-game editor such as you present, it would be ideal to disable this cache when the curves are not being edited.

As I said, it is trade-off. The price is really low(less than 1% of curve memory footprint), but simplified the code and api greatly.

> This wastes memory in the normal use case, since the cache would grow by the number of curves if there were a large number of curves. Since Godot tends to prioritize common use cases, I definitely assert that this chache should only be used in tools builds.

again, it is trade-off. Even if the `Path3DGizmo` is the sole user of  this `get_point_baked_rotation` I still believe this 1% curve memory footprint in exchange for simple code and no `_bake  performance loss is a good deal. After all Curves are not the major memory consumer for a game, far behind texture, mesh, animation track, etc.

---
## compromise 
If you are so concerned  about the memory usage of release template, but more relaxed about the tool build, then how about this.

make `_get_point_baked_rotation` a private method, but make `Path3DGizmo` a friend of the `Curve3D`. and keep all these behind the `#ifdef` gate.

Curve3D.h
```c++
class Curve3D {

#ifdef TOOLS_ENABLED
  friend Path3DGizmo;
  Vector<size_t> points_in_cache;
  Basis _get_point_baked_rotation(int idx);
#endif
```

Reply:
> I am talking about exactly the tool mode. you need additional signal an call more _bake(). That's complexity and performance loss I am talking about.

> I am talking about user who want to implement in-game-editor. When they want to get information at a control point, they have to use signal or callback function, instead of a simple method call.

You may not understand the usage of virtual methods correctly, but I can say you that these your claims are misguided.

I have more to say on the other claims, but this is all I have to say:

> Moreover, even in a use case with an in-game editor such as you present, it would be ideal to disable this cache when the curves are not being edited.

Even if the use case is to implement an editor in the game, this cache must be able to be disabled at will.

So the correct implementation for creating such a game is that adding an extend Curve3D class as a custom module, re-implementing the cache in it, and having a user property like "cache_enabled_for_editor".

> make _get_point_baked_rotation a private method, but make Path3DGizmo a friend of the Curve3D. and keep all these behind the #ifdef gate.

Avoid to use friend, you can simply put it on public but not bind it.

Reply:
> Avoid to use friend, you can simply put it on public but don't bind it.

done.

> The implementation I described above (https://github.com/godotengine/godot/pull/80329#discussion_r1290832009) should help with that.

Thanks with the advice.